docs/proposals/custom-metrics.md

```diff
@@ -0,0 +1,129 @@
+In future, we may restrict it by making the endpoints accessible only by kubelet/cAdvisor).
+CAdvisor will collect metrics from such endpoints of all pods on each node by pulling, and expose them to Heapster.
+Heapster will:
+* collect custom metrics from all CAdvisors in the cluster, together with pulling system metrics
+(for version 1.2: we assume pooling period of ~30 seconds),
+* store them in a metrics backend (influxDB, Prometheus, Hawkular, GCM, …),
+* expose the latest snapshot of custom metrics for queries (by HPA/scheduler/…) using [model API](https://github.com/kubernetes/heapster/blob/master/docs/model.md).
+
+User can easily expose Prometheus metrics for her own application by using Prometheus [client](http://prometheus.io/docs/instrumenting/clientlibs/) library.
+To monitor third-party applications, Prometheus [exporters](http://prometheus.io/docs/instrumenting/exporters/) run as side-cars containers may be used.
+
+For version 1.2, to prevent a huge number of metrics negatively affect the system performance,
+the number of metrics that can be exposed by each pod will be limited to the configurable value (default: 5).
+
+In future versions (later than 1.2), we want to extend our solution by:
+* accepting pod metrics exposed in different formats than Prometheus
+(collecting of the different formats will need to be supported by cAdvisor),
+* support push metrics by exposing push API on heapster (e.g. in StatsD format) or on a local node collector
+(if heapster performance is insufficient),
+* support metrics not associated with an individual pod.
+
+
+## API
+
+For Kubernetes 1.2, defining pod Prometheus endpoints will be done using annotations.
+Later, when we are sure that our API is correct and stable, we will make it a part of `PodSpec`.
+
+We will add a new optional pod annotation with the following key: `metrics.alpha.kubernetes.io/custom-endpoints`.
+It will contain a string-value in JSON format.
+The value will be a list of tuples defining ports, paths and protocol
+(currently, we will support only Prometheus protocol, this will be also the default value if format is empty)
+of metrics endpoints exposed by the pod, and names of metrics which should be taken
+(if empty: we will take all exposed metrics) from the endpoint.
+
+For example, the following configuration:
+
+```
+"metrics.alpha.kubernetes.io/custom-endpoints" = [
+	{
+		"protocol": "prometheus",
+		"path": "/status",
+		"port": "8080",
+		"names:" ["qps", "activeConnections"]
+	},
+	{
+		"path": "/metrics",
+		"port": "9090",
+	}
+]
+```
+
+will expose metrics with names `qps` and `activeConnections` from `localhost:8080/status` and all metrics from `localhost:9090/metrics`.
+Please note that both endpoints are in Prometheus format.
+
+
+## Implementation notes
+
+1. Kubelet will parse value of `metrics.alpha.kubernetes.io/custom-endpoints` annotation for pods.
```

Comment:
Does this imply that the annotation can not be edited post create of the pod?

When you note that in case of error, pod will not be started, will the pod just be marked failed or will it wait forever?